lib/shadow.ml

```diff
@@ -49,8 +61,8 @@ let from_shadow_t_opt = function
 
 let to_shadow_t sp =
   let sp_t : shadow_t structure = make shadow_t in
-  setf sp_t sp_name sp.name;
-  setf sp_t sp_passwd sp.passwd;
+  setf sp_t sp_name (sp.name |> string_to_char_array |> CArray.start);
+  setf sp_t sp_passwd (sp.passwd |> string_to_char_array |> CArray.start);
```

Comment:
Shouldn't we store the CArray handles somewhere to make sure they are not garbage collected? I'm not sure whether between here and the actual call there would be anything to keep the carray reference live:
`putspent' (to_shadow_t sp |> addr) fd |> ignore`
I think due to the way how CArrays are implemented this would work (`start` gives you the `astart` field, which gets stored in the struct, and taking the address of the struct hopefully keeps the struct itself alive too?).

If we want more control over lifetimes we might have to inline `to_shadow_t`, use `let sp_name_carray = ... in ... putspent`. The pipe operator is nice, but the lifetime of its input might stop too early.
To be sure maybe we could have a `let gc_id x = Gc.compact (); x` and put calls to `gc_id` on each pipe operator? That might allow us to reproduce the original problem more reliably.

Reply:
I had this doubt as well, messing around with utop I convinced myself that storing the `astart` was enough but this is far from being a certainty.

I had tests made adding `Gc.minor`, `Gc.major` and `Gc.compact` all around, and I could not reproduce the issue also with the old code unfortunately. I'll try again

Reply:
I made the lifetimes more explicit as you suggested. Stress-testing it adding various Gc actions and using memory a lot didn't show any issue (but also the original code is not failing, so it's not really a good metric)